statsmodels/robust/covariance.py

```diff
@@ -0,0 +1,867 @@
+        zs = z / scale
+        corr = cov_func(zs, scale_func=scale_func)
+        # Maronna, Zamar set diagonal to 1, otherwise small difference to 1
+        corr[np.arange(k_vars), np.arange(k_vars)] = 1
+        evals, evecs = np.linalg.eigh(corr)
+        transf = evecs * scale[:, None]   # A matrix in Maronna, Zamar
+        #z = np.linalg.solve(transf, z.T).T
+        z = zs.dot(evecs)
+        transf0 = transf0.dot(transf)
+
+    scale_z = scale_func(z)
+    cov = (transf0 * scale_z**2).dot(transf0.T)
+
+    loc_z = loc_func(z)
+    loc = transf0.dot(loc_z)
+    # prepare for results
+    cov_ogk_raw = cov
+    loc_ogk_raw = loc
+    mask = None
+    d = None
+    if reweight is not None:
+        d = (((z - loc_z) / scale_z)**2).sum(1)
+        #d = mahalanobis(x - loc, cov)
+        # only hard thresholding right now
+        dmed = np.median(d)
+        cutoff = dmed * stats.chi2.isf(1-beta, k_vars) / stats.chi2.ppf(0.5, k_vars)
+        mask = d <= cutoff
+        sample = x[mask]
+        loc = sample.mean(0)
+        cov = np.cov(sample.T, ddof=ddof)
+
+    # duplicate name loc mean center, choose consistent naming
+    res = Holder(cov=cov, loc=loc, mean=loc, mask=mask, mahalanobis=d,
+                 cov_ogk_raw=cov_ogk_raw, loc_ogk_raw=loc_ogk_raw,
+                 transf0=transf0)
+
+    return res
+
+### Tyler ###
+
+
+
+def cov_tyler(data, start_cov=None, normalize=False, maxiter=100, eps=1e-13):
+    """Tyler's M-estimator for normalized covariance (scatter)
+
+    The underlying (population) mean of the data is assumed to be zero.
+
+    Parameters
+    ----------
+    data : ndarray
+        data array with observations in rows and variables in columns
+    start_cov : None or ndarray
+        starting covariance for iterative solution
+    normalize : bool
+        If True, then the scatter matrix is normalized to have trace equalt
+        to the number of columns in the data.
+    maxiter : int
+        maximum number of iterations to find the solution
+    eps : float
+        convergence criterion. The maximum absolute distance needs to be
+        smaller than eps for convergence.
+
+    Returns
+    -------
+    result instance with the following attributes
+    cov : ndarray
+        estimate of the scatter matrix
+    iter : int
+        number of iterations used in finding a solution. If iter is less than
+        maxiter, then the iteration converged.
+    """
+    x = np.asarray(data)
+    nobs, k_vars = x.shape
+    kn = k_vars * 1. / nobs
+    if start_cov is not None:
+        c = start_cov
+    else:
+        c = np.diag(robust.mad(x, center=0)**2)
+
+    #Tyler's M-estimator of shape (scatter) matrix
+    for i in range(maxiter):
+        c_inv = np.linalg.pinv(c)
+        c_old = c
+        # this could be vectorized but could use a lot of memory
+        # TODO:  try to work in vectorized batches
+        c = kn * sum(np.outer(xi, xi) / np.inner(xi, c_inv.dot(xi)) for xi in x)
```

Comment:
Adding `from scipy.linalg.lapack import dtrtri`, this vectorized implementation is faster:
```suggestion
        c_old = c
        ichol, _ = dtrtri(linalg.cholesky(c, lower=False), lower=0)
        v = x @ ichol
        dist_mahal_2 = np.einsum('ij,ji->i', v, v.T)
        xw = np.sqrt(k_vars / dist_mahal_2[:, None]) * x
        c = xw.T @ xw / nobs
```

Reply:
Thanks
I need to check this. I would like to get to this soonish, for sure for the next release.

Does the speed increase apply in different sizes of the arrays? I don't remember what the shapes are for those arrays/matrices.

One difference between pinv and cholesky is that the former also works if `c` is singular. I don't remember if this shows up here in non-negligible case.

Reply:
I guess this will always be much faster because of the vectorization in `x` if nobs is not very small.
looks good, AFAICS

Reply:
![pr_8129_robust_cov](https://github.com/statsmodels/statsmodels/assets/5346669/100cfa2a-3dee-4305-862f-b4f1bb256092)

To reproduce this figure
```
n_obs = 100
maxiter = 10
generator = check_random_state(None)

dims = [2, 5, 7, 10, 15, 20, 30, 50]
n_dims, n_reps = len(dims), 10

t_old = np.zeros((n_dims, n_reps))
t_new = np.zeros((n_dims, n_reps))

for i, n_dim in enumerate(dims):

    data = generator.uniform(size=(n_obs, n_dim))

    for j in range(n_reps):

        t0 = timeit.default_timer()
        cov_tyler(data, maxiter=maxiter)
        t_old[i, j] = timeit.default_timer() - t0

        t0 = timeit.default_timer()
        cov_tyler_new(data, maxiter=maxiter)
        t_new[i, j] = timeit.default_timer() - t0


fig, ax = plt.subplots()
ax.errorbar(dims, t_old.mean(axis=-1), yerr=t_old.std(axis=-1), label='old')
ax.errorbar(dims, t_new.mean(axis=-1), yerr=t_new.std(axis=-1), label='new')
ax.set(xlabel='Number of variables', ylabel='Time')
plt.legend(loc='upper left')
plt.show()
```